src/main/java/org/folio/fqm/service/EntityTypeFlatteningService.java

```diff
@@ -0,0 +1,242 @@
+    List<EntityTypeSource> sources = flattenedEntityType.getSources();
+
+    // Check that exactly 1 source does not have a JOIN clause
+    long sourceWithoutJoinCount = sources
+      .stream()
+      .filter(source -> source.getJoin() == null)
+      .count();
+    if (sourceWithoutJoinCount != 1) {
+      log.error("ERROR: number of sources without joins must be exactly 1, but was {}", sourceWithoutJoinCount);
+      return ""; // TODO: handle this better
+    }
+
+    // Order sources so that JOIN clause makes sense
+    List<EntityTypeSource> orderedSources = getOrderedSources(flattenedEntityType);
+
+    for (EntityTypeSource source : orderedSources) {
+      EntityTypeSourceJoin join = source.getJoin();
+      String alias = "\"" + source.getAlias() + "\"";
+      String target = source.getTarget();
+      if (join != null) {
+        String joinClause = " " + join.getType() + " " + target + " " + alias + " ON " + join.getCondition(); // NEW
+        joinClause = joinClause.replace(":this", alias);
+        joinClause = joinClause.replace(":that", "\"" + join.getJoinTo() + "\"");
+        log.info("Join clause: " + joinClause);
+        finalJoinClause.append(joinClause);
+      } else {
+        finalJoinClause.append(target).append(" ").append(alias); // NEW
+      }
+    }
+
+    String finalJoinClauseString = finalJoinClause.toString();
+    // Replace each target in the join clause with an appropriate alias
+    for (EntityTypeSource source : sources) {
+      String toReplace = ":" + source.getAlias();
+      String alias = "\"" + source.getAlias() + "\"";
+      finalJoinClauseString = finalJoinClauseString.replace(toReplace, alias); // NEW
+    }
+    log.info("Final join clause string: " + finalJoinClauseString);
+    return finalJoinClauseString;
+  }
+
+  private List<EntityTypeSource> getOrderedSources(EntityType entityType) {
+    Map<String, EntityTypeSource> sourceMap = new HashMap<>();
+    for (EntityTypeSource source : entityType.getSources()) {
+      sourceMap.put(source.getAlias(), source);
+    }
+
+    List<EntityTypeSource> orderedList = new ArrayList<>();
+    Set<String> visited = new HashSet<>();
+
+    for (EntityTypeSource source : entityType.getSources()) {
+      if (!visited.contains(source.getAlias())) {
+        dfs(source, sourceMap, visited, orderedList);
+      }
+    }
+    return orderedList;
+  }
+
+  private static void dfs(EntityTypeSource source, Map<String, EntityTypeSource> sourceMap, Set<String> visited, List<EntityTypeSource> orderedList) {
+    visited.add(source.getAlias());
+    if (source.getJoin() != null) {
+      EntityTypeSource joinToSource = sourceMap.get(source.getJoin().getJoinTo());
+      if (!visited.contains(joinToSource.getAlias())) {
+        dfs(joinToSource, sourceMap, visited, orderedList);
+      }
+    }
+    orderedList.add(source);
+  }
+
+  private void updateOtherSources(String oldSourceName, String newSourceName, List<EntityTypeSource> otherSources) {
```

Comment:
This method name is killing me, too 😛 I dunno what "other sources" are, so it might as well be named `doThings()` 🙂